driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java

```diff
@@ -126,23 +127,43 @@ public ServerDescription getDescription() {
 
     @Override
     public void invalidate() {
+        invalidate(connectionPool.getGeneration());
+    }
+
+    @Override
+    public void invalidate(final int connectionGeneration) {
         if (!isClosed()) {
+            if (connectionGeneration < connectionPool.getGeneration()) {
```

Comment:
This is the first requirement: ignore errors from stale connections.  This particular method is called only from open/openAsync, so all we have is the connection generation.